contracts/cw-ics721-bridge/src/contract.rs

```diff
@@ -0,0 +1,451 @@
+        ExecuteMsg::DoInstantiateAndMint {
+            class_id,
+            class_uri,
+            token_ids,
+            token_uris,
+            receiver,
+        } => execute_do_instantiate_and_mint(
+            deps, env, info, class_id, class_uri, token_ids, token_uris, receiver,
+        ),
+        ExecuteMsg::ReceiveNft(cw721::Cw721ReceiveMsg {
+            sender,
+            token_id,
+            msg,
+        }) => execute_receive_nft(deps, info, token_id, sender, msg),
+        ExecuteMsg::BatchTransferFromChannel {
+            channel,
+            class_id,
+            token_ids,
+            receiver,
+        } => execute_batch_transfer_from_channel(
+            deps.as_ref(),
+            info,
+            env,
+            channel,
+            class_id,
+            token_ids,
+            receiver,
+        ),
+        ExecuteMsg::BurnEscrowTokens {
+            channel,
+            class_id,
+            token_ids,
+        } => execute_burn_escrow_tokens(deps.as_ref(), env, info, channel, class_id, token_ids),
+    }
+}
+
+fn execute_transfer(
+    deps: Deps,
+    env: Env,
+    info: MessageInfo,
+    class_id: String,
+    token_id: String,
+    receiver: String,
+) -> Result<Response, ContractError> {
+    // This will error if the class_id does not exist so no need to check
+    let owner = get_owner(deps, class_id.clone(), token_id.clone())?;
+
+    // Check if we are the owner or the contract itself
+    if info.sender != env.contract.address && info.sender != owner.owner {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let msg = transfer(deps, class_id, token_id, receiver)?;
+    Ok(Response::new()
+        .add_attribute("action", "transfer")
+        .add_submessage(msg))
+}
+
+fn execute_burn(
+    deps: Deps,
+    env: Env,
+    info: MessageInfo,
+    class_id: String,
+    token_id: String,
+) -> Result<Response, ContractError> {
+    // This will error if the class_id does not exist so no need to check
+    let owner = get_owner(deps, class_id.clone(), token_id.clone())?;
+
+    // Check if we are the owner or the contract itself
+    if info.sender != env.contract.address && info.sender != owner.owner {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let msg = burn(deps, class_id, token_id)?;
+    Ok(Response::new()
+        .add_attribute("action", "burn")
+        .add_submessage(msg))
+}
+
+fn execute_mint(
+    deps: Deps,
+    env: Env,
+    info: MessageInfo,
+    class_id: String,
+    token_ids: Vec<String>,
+    token_uris: Vec<String>,
+    receiver: String,
+) -> Result<Response, ContractError> {
+    if info.sender != env.contract.address {
+        return Err(ContractError::Unauthorized {});
+    }
+    if token_ids.len() != token_uris.len() {
+        return Err(ContractError::ImbalancedTokenInfo {});
+    }
+    let receiver = deps.api.addr_validate(&receiver)?;
+    let cw721_addr = CLASS_ID_TO_NFT_CONTRACT.load(deps.storage, class_id)?;
```

Comment:
We need to use the indexed map with two key tuple as discussed

Reply:
Per updated designs we will be using 3 different maps: 
1) CLASS_ID_TO_NFT_CONTRACT
2) NFT_CONTRACT_TO_CLASS_ID
3) NFT_TRACKING_MAP: (channel, class_id, token_id): <contract address>